### ISSUE TITLE
About the results of the Actinomock Example

### ISSUE THREAD

**Issue by WayneWu01** [Lizhen0909/LSHVec#5]
I have a question of how to evaluate the result of the example. I read your paper, but don't know how to find those FN things. Could you elaborate me more?

**Comment by Lizhen0909** [comment]
You may find examples in 'notebook' folder (The data link in the notebook may not exist anymore. But it can be download from https://www.amazon.com/clouddrive/share/eTIKYVLckXUCMnMQSpO8TCqZOwekmBrx23ZhMa3XO8d).

Also a python wrapper of pretrained models can be found at https://github.com/Lizhen0909/pyLSHVec.

**Comment by WayneWu01** [comment]
I’m able to reproduce your example, But I was wondering how to analyze your
result? Like how to know whether the example works good or not.

Lizhen Shi ***@***.***>于2022年7月1日 周五上午10:12写道：

**Comment by Lizhen0909** [comment]
There are two tasks: embedding and classification.
For the actinomock example data the second columun looks like '47914-2616644829-Gammaproteobacteria-Proteobacteria'.
Here either of Proteobacteria, Gammaproteobacteria, 2616644829 can be the label of this row (depending on which level do you want on the taxonomy hierarchy). 

########
Embedding is and unsupervised model to get embedding vectors for kmers/sequences. 
Since it is unsupervised, there is no numerical metrics.  
Also because the dimension is high (e.g. 100), we have to use tools like tsne or umap to visualize the vectors in low dimension (e.g. 2).  
Then with the prior knowledge of the sequences (a.k.a labels of sequence, which is not used in training), we can judge if the vectors is good or not  (It is expected that vectors of same labels are clustered together)
(You may find visualization examples of language embedding at https://projector.tensorflow.org/)

###########
Classification is just exactly  same as what it is in machine learning.
Basically data is spitted as training and test datasets. Model is trained on training dataset. Metrics are reported on test dataset.
All metrics (accuracy, precision, recall, F1, AUC) applied to multiclass classification problems can be used .  I did not get what FN is earlier.   I think you mean false negative. It is part of the definition of precision and recall.

**Comment by WayneWu01** [comment]
I could get your result for the tsne part, but I have question on how to
find the number of false negative and all those numbers. Are you training
model for example use 80 percent coverage for the sequence ? How do you use
the mod.bin for the predicting of the rest 20 percent?

Lizhen Shi ***@***.***>于2022年7月1日 周五下午5:08写道：

**Comment by Lizhen0909** [comment]
use command line:
lshvec predict  <model.bin>  <test_data_file>
or 
lshvec predict-prob<model.bin>  <test_data_file>

**Comment by WayneWu01** [comment]
Thank you so much！I will try it!

Lizhen Shi ***@***.***>于2022年7月1日 周五下午11:04写道：

**Comment by WayneWu01** [comment]
Just for confirm, about the input of test data file. Which format should I
use? The seq file or the hash file?

Zheng Wu ***@***.***>于2022年7月1日 周五下午11:06写道：

**Comment by Lizhen0909** [comment]
Hashed data. 
It should be same as the training data.

**Comment by WayneWu01** [comment]
When I used the predict, it gave me "terminate called after throwing an
instance of 'std::invalid_argument' what(): stoi". I use the command in
following structure:
             .../lshvec predict  \
            --model_path .../model.bin \
            --hash_path .../test.hash
What is the correct way of using the predict command? I could use
lshvec skipgram for training.

**Comment by WayneWu01** [comment]
Could you help me with the command?

Zheng Wu ***@***.***>于2022年7月6日 周三下午8:09写道：

**Comment by Lizhen0909** [comment]
Should be something like this:
>  .../lshvec predict   .../model.bin   .../test.hash

Here predict or predict-prob is subcommand, you can find them at src/main_fastseq.cc
If you run command without arguments, it should print usages.
For example:
> lshvec  #print all subcommands
> lshvec predict #print usage for predict sub commands

**Comment by WayneWu01** [comment]
I tried your way and it gave me all the details, but it still gave me the
error of "Model needs to be supervised for prediction!". I read that the k
and threshold are optional, why did it still give me this error? So sorry
for bothering you again and again!

Zheng Wu ***@***.***>于2022年7月8日 周五下午1:49写道：

**Comment by WayneWu01** [comment]
I tried exactly your way for prediction. It still gave me the error that
the "Model needs to be supervised for prediction!" How did you predict your
model? Sorry for bothering you again.

**Comment by Lizhen0909** [comment]
It seems that your model was trained for embedding but for classification.
You may refer to https://github.com/Lizhen0909/LSHVec/blob/master/notebook/lsa_spike_fnv_classfication.ipynb, which is a bit older.  
Notice the cell 35 where uses subcommand "supervised"

**Comment by WayneWu01** [comment]
I see, then how to test the accuracy or the performance of the
skipgram model? What is the standard?

**Comment by Lizhen0909** [comment]
From 'predict' subcommand you get predictions.  Then it depends on you to evaluate the performance, no standard.  
For example, comparing the predictions to the ground truth label to get accuracy, precision and recall.

**Comment by WayneWu01** [comment]
But how to evaluate the skipgram model? Which model does supervised command
use?

Lizhen Shi ***@***.***>于2022年7月13日 周三下午8:53写道：

**Comment by WayneWu01** [comment]
Is the following process right? Separate the hash file by 1:4. Train the
model use the 4 part and get the prediction of 1 part. Then compare it with
the seq file which is the ground truth to get accuracy.

Zheng Wu ***@***.***>于2022年7月3日 周日上午5:27写道：

**Comment by WayneWu01** [comment]
I tried your way and it gave me all the details, but it still gave me the
error of "Model needs to be supervised for prediction!". I read that the k
and threshold are optional, why did it still give me this error? So sorry
for bothering you again and again!

**Comment by WayneWu01** [comment]
How did you get this table? Just use the supervised model for different
types of hashing methods? Which model did the supervised use? Just
this " std::uniform_int_distribution<>
uniform(0, labels.size() - 1); "?
